icicle/backend/cpu/include/ntt_utils.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+#include "icicle/fields/field_config.h"
+
+// #include <_types/_uint32_t.h>
+#include <sys/types.h>
+
+using namespace field_config;
+using namespace icicle;
+namespace ntt_cpu {
+
+  /**
+   * @brief Represents the coordinates of a task in the NTT hierarchy.
+   * This struct holds indices that identify the position of a task within the NTT computation hierarchy.
+   *
+   * @param hierarchy_1_layer_idx Index of the hierarchy_1 layer.
+   * @param hierarchy_1_subntt_idx Index of the sub-NTT within the hierarchy_1 layer.
+   * @param hierarchy_0_layer_idx Index of the hierarchy_0 layer.
+   * @param hierarchy_0_block_idx Index of the block within the hierarchy_0 layer.
+   * @param hierarchy_0_subntt_idx Index of the sub-NTT within the hierarchy_0 block.
+   *
+   * @method bool operator==(const NttTaskCoordinates& other) const Compares two task coordinates for equality.
+   */
+  struct NttTaskCoordinates {
+    uint32_t hierarchy_1_layer_idx = 0;
+    uint32_t hierarchy_1_subntt_idx = 0;
+    uint32_t hierarchy_0_layer_idx = 0;
+    uint32_t hierarchy_0_block_idx = 0;
+    uint32_t hierarchy_0_subntt_idx = 0;
+    bool reorder = false;
+
+    bool operator==(const NttTaskCoordinates& other) const
+    {
+      return hierarchy_1_layer_idx == other.hierarchy_1_layer_idx &&
+             hierarchy_1_subntt_idx == other.hierarchy_1_subntt_idx &&
+             hierarchy_0_layer_idx == other.hierarchy_0_layer_idx &&
+             hierarchy_0_block_idx == other.hierarchy_0_block_idx &&
+             hierarchy_0_subntt_idx == other.hierarchy_0_subntt_idx;
+    }
+  };
+
+  uint64_t bit_reverse(uint64_t i, uint32_t logn)
+  {
+    uint32_t rev = 0;
+    for (uint32_t j = 0; j < logn; ++j) {
```

Comment:
maybe the following code will be more efficient
      rev_idx = ((rev_idx >> 1) & 0x5555555555555555) | ((rev_idx & 0x5555555555555555) << 1); // bit rev single bits
      rev_idx = ((rev_idx >> 2) & 0x3333333333333333) | ((rev_idx & 0x3333333333333333) << 2); // bit rev 2 bits chunk
      rev_idx = ((rev_idx >> 4) & 0x0F0F0F0F0F0F0F0F) | ((rev_idx & 0x0F0F0F0F0F0F0F0F) << 4); // bit rev 4 bits chunk
      rev_idx = ((rev_idx >> 8) & 0x00FF00FF00FF00FF) | ((rev_idx & 0x00FF00FF00FF00FF) << 8); // bit rev 8 bits chunk
      rev_idx =
        ((rev_idx >> 16) & 0x0000FFFF0000FFFF) | ((rev_idx & 0x0000FFFF0000FFFF) << 16); // bit rev 16 bits chunk
      rev_idx = (rev_idx >> 32) | (rev_idx << 32);                                       // bit rev 32 bits chunk
      rev_idx = rev_idx >> (64 - m_bit_size);                                            // Align rev_idx to the LSB